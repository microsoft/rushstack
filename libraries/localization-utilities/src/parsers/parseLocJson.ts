// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { JsonFile, JsonSchema } from '@rushstack/node-core-library';

import type { ILocalizationFile, IParseFileOptions } from '../interfaces';
import locJsonSchema from '../schemas/locJson.schema.json';

const LOC_JSON_SCHEMA: JsonSchema = JsonSchema.fromLoadedObject(locJsonSchema);

/**
 * @public
 */
export function parseLocJson({ content, filePath, ignoreString }: IParseFileOptions): ILocalizationFile {
  const parsedFile: ILocalizationFile = JsonFile.parseString(content);
  try {
    LOC_JSON_SCHEMA.validateObject(parsedFile, filePath, { ignoreSchemaField: true });
  } catch (e) {
    throw new Error(`The loc file is invalid. Error: ${e}`);
  }

  if (ignoreString) {
    const newParsedFile: ILocalizationFile = {};
    for (const [key, stringData] of Object.entries(parsedFile)) {
      if (!ignoreString(filePath, key)) {
        newParsedFile[key] = stringData;
      }
    }

    return newParsedFile;
  } else {
    return parsedFile;
  }
}
