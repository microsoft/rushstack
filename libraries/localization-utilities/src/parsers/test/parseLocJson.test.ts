// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { IgnoreStringFunction } from '../../interfaces';
import { parseLocJson } from '../parseLocJson';

describe(parseLocJson.name, () => {
  it('parses a valid file', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        comment: 'A string'
      },
      bar: {
        value: 'Bar',
        comment: 'Another string'
      }
    });

    expect(
      parseLocJson({
        content,
        filePath: 'test.loc.json'
      })
    ).toMatchSnapshot();
  });

  it('parses a file with raw strings', () => {
    const content: string = JSON.stringify({
      foo: 'Foo',
      bar: 'Bar'
    });

    expect(
      parseLocJson({
        content,
        filePath: 'test.loc.json'
      })
    ).toMatchSnapshot();
  });

  it('throws on invalid file', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        baz: 'A string'
      }
    });

    expect(() =>
      parseLocJson({
        content,
        filePath: 'test.loc.json'
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it('correctly ignores a string', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        comment: 'A string'
      },
      bar: {
        value: 'Bar',
        comment: 'Another string'
      }
    });

    const ignoredStringFunction: IgnoreStringFunction = jest
      .fn()
      .mockImplementation(
        (fileName: string, stringName: string) => fileName === 'test.loc.json' && stringName === 'bar'
      );

    expect(
      parseLocJson({
        content,
        filePath: 'test.loc.json',
        ignoreString: ignoredStringFunction
      })
    ).toMatchSnapshot('Loc file');

    expect((ignoredStringFunction as unknown as jest.SpyInstance).mock.calls).toMatchSnapshot(
      'ignoreStrings calls'
    );
  });
});
