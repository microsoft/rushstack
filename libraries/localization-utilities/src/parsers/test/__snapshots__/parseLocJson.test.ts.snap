// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseLocJson correctly ignores a string: Loc file 1`] = `
Object {
  "foo": Object {
    "comment": "A string",
    "value": "Foo",
  },
}
`;

exports[`parseLocJson correctly ignores a string: ignoreStrings calls 1`] = `
Array [
  Array [
    "test.loc.json",
    "foo",
  ],
  Array [
    "test.loc.json",
    "bar",
  ],
]
`;

exports[`parseLocJson parses a file with raw strings 1`] = `
Object {
  "bar": Object {
    "value": "Bar",
  },
  "foo": Object {
    "value": "Foo",
  },
}
`;

exports[`parseLocJson parses a valid file 1`] = `
Object {
  "bar": Object {
    "comment": "Another string",
    "value": "Bar",
  },
  "foo": Object {
    "comment": "A string",
    "value": "Foo",
  },
}
`;

exports[`parseLocJson throws on invalid file 1`] = `
"The loc file is invalid. Error: Error: JSON validation failed:
test.loc.json

Error: #/foo
       must NOT have additional properties: baz
Error: #/foo
       must be string
Error: #/foo
       must match exactly one schema in oneOf"
`;
